R/openmp.R

```diff
@@ -0,0 +1,146 @@
+##' Return information about OpenMP support for this system.  For
+##' individual models look at the `$has_openmp()` method
+##'
+##' @title Information about OpenMP support
+##'
+##' @param check_compile Logical, indicating if we should check if we
+##'   can compile an openmp program - this is slow the first time.
+##'
+##' @seealso [dust_openmp_threads()] for setting a polite number of
+##'   threads.
+##'
+##' @export
+##' @examples
+##' # System wide support
+##' dust::dust_openmp_support()
+##'
+##' # Support compiled into a generator
+##' walk <- dust::dust_example("walk")
+##' walk$public_methods$has_openmp()
+##'
+##' # Support from an instance of that model
+##' model <- walk$new(list(sd = 1), 0, 1)
+##' model$has_openmp()
+dust_openmp_support <- function(check_compile = FALSE) {
+  info <- openmp_info()
+  if (check_compile) {
+    info$has_openmp_compiler <- has_openmp_compiler()
+  }
+  info
+}
+
+
+##' Politely select a number of threads to use. See Details for the algorithm
+##'
+##' There are two limits and we will take the smaller of the two.
+##'
+##' The first limit comes from piggy-backing off of R's normal
+##' parallel configuration; we will use the `MC_CORES` environment
+##' variable and `mc.cores` option as a guide to how many cores you
+##' are happy to use. We take `mc.cores` first, then `MC_CORES`, which
+##' is the same behaviour as `parallel::mclapply` and friends.
+##'
+##' The second limit comes from openmp. If you do not have OpenMP
+##' support, then we use one thread (higher numbers have no effect at
+##' all in this case). If you do have OpenMP support, we take the
+##' smallest of the number of "processors" (reported by
+##' `omp_get_num_procs()`) the "max threads" (reported by
+##' `omp_get_max_threads()` and "thread_limit" (reported by
+##' `omp_get_thread_limit()`.
+##'
+##' See [dust::dust_openmp_support()] for the values of all the values
+##' that go into this calculation.
+##'
+##' @title Select number of threads
+##'
+##' @param n Either `NULL` (select automatically) or an integer as
+##'   your proposed number of threads.
+##'
+##' @param action An action to perform if `n` exceeds the maximum
+##'   number of threads you can use. Options are "error" (the default,
+##'   throw an error), "fix" (print a message and reduce `n` down to
+##'   the limit) or "message" (print a message and continue anyway)
```

Comment:
This is a very 'feature rich' implementation. Not a problem, but I think many parallel implementations (e.g. openmp itself) would silently do the 'fix' behaviour.

Reply:
It will for the openmp part of the number, but not the R part, and I was thinking a bit of a message might be useful as it'd end up in a log (why did my thing not run any faster? I swear I increased the number of threads). And the CRAN issue looms as there are some fun issues over on data.table where they were told off by CRAN for not honouring some of these openmp limits